src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,411 @@
+            # overridden to reduce memory usage when there's a large backlog
+            "queued.max.messages.kbytes": DEFAULT_QUEUED_MAX_MESSAGE_KBYTES,
+            "queued.min.messages": DEFAULT_QUEUED_MIN_MESSAGES,
+        },
+    )
+    return consumer_config
+
+
+class MetricsBatchBuilder:
+    """
+    Batches up individual messages - type: Message[KafkaPayload] - into a
+    list, which will later be the payload for the big outer message
+    that gets passed through to the ParallelTransformStep.
+
+    See `__flush` method of BatchMessages for when that happens.
+    """
+
+    def __init__(self, max_batch_size: int, max_batch_time: float) -> None:
+        self.__messages: MessageBatch = []
+        self.__max_batch_size = max_batch_size
+        self.__deadline = time.time() + max_batch_time
+
+    def __len__(self) -> int:
+        return len(self.__messages)
+
+    @property
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[KafkaPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep[KafkaPayload]):
+    """
+    First processing step in the MetricsConsumerStrategyFactory.
+    Keeps track of a batch of messages (using the MetricsBatchBuilder)
+    and then when at capacity, either max_batch_time or max_batch_size,
+    flushes the batch.
+
+    Flushing the batch here means wrapping the batch in a Message, the batch
+    itself being the payload. This is what the ParallelTransformStep will
+    process in the process_message function.
+    """
+
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[MessageBatch],
+        max_batch_time: float,
+        max_batch_size: int,
+    ):
+        self.__max_batch_size = max_batch_size
+        self.__max_batch_time = max_batch_time
+
+        self.__next_step = next_step
+        self.__batch: Optional[MetricsBatchBuilder] = None
+        self.__closed = False
+
+    def poll(self) -> None:
+        assert not self.__closed
+
+        self.__next_step.poll()
+
+        if self.__batch and self.__batch.ready():
+            self.__flush()
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        if self.__batch is None:
+            self.__batch = MetricsBatchBuilder(self.__max_batch_size, self.__max_batch_time)
+
+        self.__batch.append(message)
+
+    def __flush(self) -> None:
+        if not self.__batch:
+            return
+        last = self.__batch.messages[-1]
+
+        new_message = Message(last.partition, last.offset, self.__batch.messages, last.timestamp)
+
+        self.__next_step.submit(new_message)
+        self.__batch = None
+
+    def terminate(self) -> None:
+        self.__closed = True
```

Comment:
please propagate the terminate call to the next step 
`self.__next_step.terminate()`